### PR TITLE
Update pythia8 pt hat reweight user hook

### PR DESCRIPTION
**Pull request by aperloff**
This PR would provide a new UserHook to Pythia in order to generate a sample with a truly "flat" pThat spectrum. These types of samples are currently used by the JetMET group for jet energy calibrations, but currently we are using a sub-optimal configuration; the new method works much better. It is not foreseen that this PR will have any effect on other packages. This should simply add a new feature to the CMSSW-Pythia8 interface.

We would like to include this feature in CMSSW in time for the upcoming 2017 MC production campaign using CMSSW_9_3_0.

### CODE REVIEW

**Comment by perrozzi** [comment]
please test

**Comment by aperloff** [comment]
I have a couple of questions after the failed tests:

1) Why was an additional PR (0619998) added on top of the merge for this PR? I just don't know what this has to do with the current PR.
2) The AddOn test that failed seem to be looking for the file MagneticField/GeomBuilder/data/MagneticFieldVolumes_1_v7_large.xml, which isn't in the search path. I didn't touch that file with this PR, so is it missing in the IB? Is it something I did wrong?

**Comment by aperloff** [comment]
Tagging some people involved in this PR who might be interested in its progress
@karavdin, @kirschen, @miquork

**Comment by aperloff** [comment]
Have the AddOn tests been updated or has the IB been fixed? Can the tests for this PR be retried or will they just fail again?

**Comment by kpedro88** [comment]
please test

**Comment by perrozzi** [comment]
+1

File: GeneratorInterface/Pythia8Interface/plugins/ReweightUserHooks.h [hunk 1]
```diff
@@ -26,6 +26,34 @@ class PtHatReweightUserHook : public Pythia8::UserHooks
     double pt, power;
 };
 
+class PtHatEmpReweightUserHook : public Pythia8::UserHooks
+{
+  public:
+   PtHatEmpReweightUserHook(double _pTHatMin, double _pTHatMax) :
+      pTHatMin(_pTHatMin), pTHatMax(_pTHatMax) {
+      sigma = TF1("sigma", "([0]*pow(x,[2]+[3]*log(0.01*x)+[4]*pow(log(0.01*x),2))*pow(1-2*x/(13000.+[5]),[1]))*x", pTHatMin, pTHatMax);
```
Comment (davidlange6) [inline GeneratorInterface/Pythia8Interface/plugins/ReweightUserHooks.h]: hi @aperloff - please just make this a C++ function. Everything is hardwired anyway, so no need for the overhead of ROOT's TF1.

**Comment by aperloff** [comment]
@davidlange6 I've updated the code to remove the TF1. Would you mind taking a look and letting me know if this is okay? I've already tested the code to make sure it returns the same values as the previous iteration, which it does. If it meets with your approval I think it is ready for testing again.

**Comment by perrozzi** [comment]
please test

**Comment by aperloff** [comment]
@perrozzi If this PR meets with your approval would you please sign off on it before this week's ORP meeting? I just want to make sure the PR can be approved in time for CMSSW_9_3_0.

**Comment by perrozzi** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Comment by Dr15Jones** [comment]
This pull request is causing failures in the IB RelVals:

https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_amd64_gcc630/CMSSW_9_3_X_2017-07-20-1100/pyRelValMatrixLogs/run/5.6_QCD_FlatPt_15_3000+QCDFlatPt153000FS+HARVESTFS/step1_QCD_FlatPt_15_3000+QCDFlatPt153000FS+HARVESTFS.log

**Comment by aperloff** [comment]
@Dr15Jones This PR added two parameters to an existing function (https://github.com/cms-sw/cmssw/pull/19668/files#diff-96dae42223a6b851894a2bd75bc7314dR219). Were there IB RelVal tests that used the PtHatReweightUserHook class (https://github.com/cms-sw/cmssw/pull/19668/files#diff-96dae42223a6b851894a2bd75bc7314dL213)? Why did they not show up when doing runTheMatrix?

**Comment by Dr15Jones** [comment]
The pull request tests only run a very small number of potential workflows from runTheMatrix, and apparently the flat distribution workflow is not used. The IB runs all the workflows.

**Comment by aperloff** [comment]
@Dr15Jones Can the test be updated or can I do another PR to make the parameter optional? If the latter, how would I go about doing that in a way that is acceptable to the developers of CMSSW? I was thinking I could use something like:
if(params.exists( "pTRef" ) && params.exists( "power" ))
    constructor with parameters
else
    constructor without parameters

**Comment by Dr15Jones** [comment]
You'd need to create a new pull request. The use of `ParameterSet::exists` is being deprecated in CMSSW since the use of `fillDescriptions` is preferred.

Rather than adding more `exists` (which is ultimately up to the generator L2 to decide) I would suggest adding in the newly needed parameters into the configuration files used by the workflows.

**Comment by aperloff** [comment]
Where are the configuration files used by the workflows stored? I've never edited a runTheMatrix workflow. I'd be happy to do the change there as that would be the easiest path.

**Comment by Dr15Jones** [comment]
I'd suggest doing a github search on the parameter used to turn on the flat distribution. I would expect that to find the files that need to be updated.

**Comment by Dr15Jones** [comment]
It looks like they are kept in Configuration/Generator/python.

**Comment by Dr15Jones** [comment]
Doing a `git grep reweightGen`  gave the following hits in python

```
Configuration/Generator/python/InclusiveppMuX_8TeV_TuneCUETP8M1_cfi.py:                         reweightGen = cms.PSet(),
Configuration/Generator/python/QCDForPF_13TeV_TuneCUETP8M1_cfi.py:                         reweightGen = cms.PSet(),
Configuration/Generator/python/QCDForPF_14TeV_TuneCUETP8M1_cfi.py:                         reweightGen = cms.PSet(),
Configuration/Generator/python/QCDForPF_8TeV_TuneCUETP8M1_cfi.py:                         reweightGen = cms.PSet(), 
Configuration/Generator/python/QCD_Pt-15To7000_TuneCUETP8M1_Flat_14TeV-pythia8_cff.py:        #reweightGen = cms.bool(True), #
GeneratorInterface/Pythia8Interface/test/pythia8ex2_cfg.py:    #reweightGen = cms.PSet(),
GeneratorInterface/Pythia8Interface/test/pythia8ex2_cfg.py:    #reweightGenRap = cms.PSet( # flat in eta
GeneratorInterface/Pythia8Interface/test/pythia8ex2_cfg.py:    #reweightGenPtHatRap = cms.PSet( # flat in Pt and eta
```

**Comment by aperloff** [comment]
I'm working on a new PR right now.

**Comment by aperloff** [comment]
A new PR #19843 has been made which should fix this problem.

**Comment by alberto-sanchez** [comment]
@aperloff, can you please do the PR for backporting this to 92x and eventually 71x?

**Pull request merged**